src/h3/mod.rs

```diff
@@ -26,4 +26,778 @@
+
+        let req_frame = frame::Frame::Headers { header_block };
+
+        let mut b = octets::Octets::with_slice(&mut d);
+        req_frame.to_bytes(&mut b)?;
+
+        let stream_id = self.get_available_request_stream()?;
+        self.streams
+            .insert(stream_id, stream::Stream::new(stream_id, !has_body));
+
+        let off = b.off();
+
+        trace!(
+            "{} sending request of size {} on stream {}",
+            quic_conn.trace_id(),
+            off,
+            stream_id
+        );
+
+        if let Err(e) = quic_conn.stream_send(stream_id, &d[..off], true) {
+            error!("{} stream send failed {:?}", quic_conn.trace_id(), e);
+            return Err(Error::from(e));
+        }
+
+        Ok(stream_id)
+    }
+
+    /// Prepare a response in HTTP/3 wire format and send it.
+    pub fn send_response(
+        &mut self, quic_conn: &mut super::Connection, stream_id: u64,
+        headers: &[qpack::Header], has_body: bool,
+    ) -> Result<()> {
+        let mut stream_d = [42; 65535];
+
+        let headers_len = headers
+            .iter()
+            .fold(0, |acc, h| acc + h.value().len() + h.name().len() + 32);
+        let mut header_block = vec![0; headers_len];
+        let len = self.qpack_encoder.encode(&headers, &mut header_block)?;
+        header_block.truncate(len);
+
+        let resp_frame = frame::Frame::Headers { header_block };
+
+        let mut stream_b = octets::Octets::with_slice(&mut stream_d);
+        resp_frame.to_bytes(&mut stream_b).unwrap();
+
+        let off = stream_b.off();
+
+        trace!(
+            "{} sending response HEADERS frame of size {} on stream {}",
+            quic_conn.trace_id(),
+            off,
+            stream_id
+        );
+
+        quic_conn.stream_send(stream_id, &stream_d[..off], !has_body)?;
+
+        Ok(())
+    }
+
+    /// Prepare paload body in HTTP/3 wire format and send it.
+    pub fn send_body(
+        &mut self, quic_conn: &mut super::Connection, stream_id: u64,
+        body: Vec<u8>, final_data: bool,
+    ) -> Result<()> {
+        let mut stream_d = [42; 65535];
+
+        let data = frame::Frame::Data { payload: body };
+
+        let mut stream_b = octets::Octets::with_slice(&mut stream_d);
+        data.to_bytes(&mut stream_b).unwrap();
+
+        let off = stream_b.off();
+
+        trace!(
+            "{} sending response DATA frame of size {} on stream {}",
+            quic_conn.trace_id(),
+            off,
+            stream_id
+        );
+
+        quic_conn.stream_send(stream_id, &stream_d[..off], final_data)?;
+
+        Ok(())
+    }
+
+    /// Read QUIC streams, assign HTTP/3 meaning to them and process HTTP/3
+    /// frames
+    pub fn process(
+        &mut self, quic_conn: &mut super::Connection,
+    ) -> Result<(u64, Event)> {
+        // Read streams and assign HTTP/3 meaning to them.
+        let streams: Vec<u64> = quic_conn.readable().collect();
+
+        for s in streams {
```

Comment:
You might be able to just do:

```suggestion
        for s in quic_conn.readable() {
```

so you don't need the `collect()` above.